File: oma-pkg/src/oma/providers/http_providers.py
"""
OMA HTTP Providers -- raw HTTP connectors for each subscription.

No SDK dependencies. Each provider is defined as:
  - endpoint URL
  - default model
  - headers builder
  - request body builder
  - response parser

This is the side-channel: we talk to every provider the same way,
using their public HTTP APIs directly.
"""

import json
import time
import urllib.error
import urllib.request
from collections.abc import Callable
from typing import Any, TypedDict

from .base import Provider, ProviderResponse


class HTTPProvider(Provider):
    """Generic HTTP-based LLM provider."""

    def __init__(
        self,
        name: str,
        api_key: str,
        endpoint: str,
        model: str,
        headers_fn: Callable,
        body_fn: Callable,
        parse_fn: Callable,
        **kwargs,
    ):
        super().__init__(api_key=api_key, **kwargs)
        self.name = name
        self.endpoint = endpoint
        self.model = model
        self._headers_fn = headers_fn
        self._body_fn = body_fn
        self._parse_fn = parse_fn

    def complete(
        self,
        messages: list[dict],
        system: str | None = None,
        max_tokens: int = 4096,
        temperature: float = 0.3,
        **kwargs,
    ) -> ProviderResponse:
        self.rate_limiter.wait_if_needed(estimated_tokens=max_tokens)

        headers = self._headers_fn(self.api_key)
        model = kwargs.get("model", self.model)
        body = self._body_fn(
            messages=messages,
            system=system,
            model=model,
            max_tokens=max_tokens,
            temperature=temperature,
        )

        endpoint = self.endpoint
        if "{model}" in endpoint:
            endpoint = endpoint.replace("{model}", model)
        if self.name == "gemini" and self.api_key and "key=" not in endpoint:
            separator = "&" if "?" in endpoint else "?"
            endpoint = f"{endpoint}{separator}key={self.api_key}"

        t0 = time.time()
        try:
            req = urllib.request.Request(
                endpoint,
                data=json.dumps(body).encode("utf-8"),
                headers=headers,
                method="POST",
            )
            with urllib.request.urlopen(req, timeout=self.timeout) as resp:
                raw = json.loads(resp.read().decode("utf-8"))

            latency = (time.time() - t0) * 1000
            result = self._parse_fn(raw)
            self.rate_limiter.record_tokens(result.tokens_total)
            self.on_success()

            return ProviderResponse(
                text=result.text,
                tokens_in=result.tokens_in,
                tokens_out=result.tokens_out,
                model=result.model or self.model,
                provider=self.name,
                latency_ms=latency,
                raw=raw,
            )

        except urllib.error.HTTPError as e:
            latency = (time.time() - t0) * 1000
            error_body = e.read().decode("utf-8", errors="replace") if e.fp else str(e)
            return ProviderResponse(
                text="",
                tokens_in=0,
                tokens_out=0,
                model=self.model,
                provider=self.name,
                latency_ms=latency,
                error=f"HTTP {e.code}: {error_body[:500]}",
                error_class=self.classify_error(e),
            )

        except Exception as e:
            latency = (time.time() - t0) * 1000
            return ProviderResponse(
                text="",
                tokens_in=0,
                tokens_out=0,
                model=self.model,
                provider=self.name,
                latency_ms=latency,
                error=str(e),
                error_class=self.classify_error(e),
            )

    def count_tokens(self, text: str) -> int:
        # rough estimate: 1 token ~ 4 chars for most models
        return len(text) // 4


class _ParsedResponse:
    """Intermediate parsed response."""
    def __init__(self, text="", tokens_in=0, tokens_out=0, model=None):
        self.text = text
        self.tokens_in = tokens_in
        self.tokens_out = tokens_out
        self.model = model
        self.tokens_total = tokens_in + tokens_out


# ---- provider configs ----

def _claude_headers(api_key):
    return {
        "Content-Type": "application/json",
        "x-api-key": api_key,
        "anthropic-version": "2023-06-01",
    }

def _claude_body(messages, system, model, max_tokens, temperature):
    body = {
        "model": model,
        "max_tokens": max_tokens,
        "temperature": temperature,
        "messages": messages,
    }
    if system:
        body["system"] = system
    return body

def _claude_parse(raw):
    content = raw.get("content", [{}])
    text = content[0].get("text", "") if content else ""
    usage = raw.get("usage", {})
    return _ParsedResponse(
        text=text,
        tokens_in=usage.get("input_tokens", 0),
        tokens_out=usage.get("output_tokens", 0),
        model=raw.get("model"),
    )


def _openai_headers(api_key):
    return {
        "Content-Type": "application/json",
        "Authorization": f"Bearer {api_key}",
    }

def _openai_body(messages, system, model, max_tokens, temperature):
    msgs = list(messages)
    if system:
        msgs.insert(0, {"role": "system", "content": system})
    return {
        "model": model,
        "max_tokens": max_tokens,
        "temperature": temperature,
        "messages": msgs,
    }

def _openai_parse(raw):
    choices = raw.get("choices", [{}])
    text = choices[0].get("message", {}).get("content", "") if choices else ""
    usage = raw.get("usage", {})
    return _ParsedResponse(
        text=text,
        tokens_in=usage.get("prompt_tokens", 0),
        tokens_out=usage.get("completion_tokens", 0),
        model=raw.get("model"),
    )


def _openai_responses_body(messages, system, model, max_tokens, temperature):
    """Build a stateless request for OpenAI's Responses API.

    ``temperature`` is intentionally not sent: current GPT reasoning models
    reject it unless reasoning is disabled.  OMA owns conversation state, so
    provider-side storage is also disabled explicitly.
    """
    body = {
        "model": model,
        "input": list(messages),
        "max_output_tokens": max_tokens,
        "store": False,
    }
    if system:
        body["instructions"] = system
    return body


def _openai_responses_parse(raw):
    """Extract text and usage from a non-streaming Responses API result."""
    text_parts = []
    for item in raw.get("output", []):
        if item.get("type") != "message":
            continue
        for content in item.get("content", []):
            if content.get("type") == "output_text" and content.get("text"):
                text_parts.append(content["text"])

    usage = raw.get("usage", {})
    return _ParsedResponse(
        text="".join(text_parts),
        tokens_in=usage.get("input_tokens", 0),
        tokens_out=usage.get("output_tokens", 0),
        model=raw.get("model"),
    )


def _gemini_headers(api_key):
    return {"Content-Type": "application/json"}

def _gemini_body(messages, system, model, max_tokens, temperature):
    contents = []
    for m in messages:
        role = "user" if m["role"] == "user" else "model"
        contents.append({"role": role, "parts": [{"text": m["content"]}]})

    body = {
        "contents": contents,
        "generationConfig": {
            "maxOutputTokens": max_tokens,
            "temperature": temperature,
        },
    }
    if system:
        body["systemInstruction"] = {"parts": [{"text": system}]}
    return body

def _gemini_parse(raw):
    candidates = raw.get("candidates", [{}])
    parts = candidates[0].get("content", {}).get("parts", []) if candidates else []
    text = parts[0].get("text", "") if parts else ""
    usage = raw.get("usageMetadata", {})
    return _ParsedResponse(
        text=text,
        tokens_in=usage.get("promptTokenCount", 0),
        tokens_out=usage.get("candidatesTokenCount", 0),
    )


def _deepseek_headers(api_key):
    return _openai_headers(api_key)  # openai-compatible

def _deepseek_body(messages, system, model, max_tokens, temperature):
    return _openai_body(messages, system, model, max_tokens, temperature)

def _deepseek_parse(raw):
    return _openai_parse(raw)  # openai-compatible


def _optional_bearer_headers(api_key):
    h = {"Content-Type": "application/json"}
    if api_key:
        h["Authorization"] = f"Bearer {api_key}"
    return h


# Jev, Groq, Mistral, Ollama, OpenRouter, Together, Qwen, GLM, and Kimi use OpenAI-compatible APIs

class ProviderConfig(TypedDict):
    """Everything needed to talk to one provider over raw HTTP."""

    endpoint: str
    default_model: str
    headers_fn: Callable[..., Any]
    body_fn: Callable[..., Any]
    parse_fn: Callable[..., Any]


PROVIDER_CONFIGS: dict[str, ProviderConfig] = {
    "claude": {
        "endpoint": "https://api.anthropic.com/v1/messages",
        "default_model": "claude-sonnet-4-20250514",
        "headers_fn": _claude_headers,
        "body_fn": _claude_body,
        "parse_fn": _claude_parse,
    },
    "chatgpt": {
        # OpenAI recommends Responses for new integrations.  The other
        # OpenAI-compatible providers below intentionally retain their Chat
        # Completions protocol.
        "endpoint": "https://api.openai.com/v1/responses",
        "default_model": "gpt-5.4",
        "headers_fn": _openai_headers,
        "body_fn": _openai_responses_body,
        "parse_fn": _openai_responses_parse,
    },
    "gemini": {
        # api key appended as query param at call time
        "endpoint": "https://generativelanguage.googleapis.com/v1beta/models/{model}:generateContent",
        "default_model": "gemini-2.0-flash",
        "headers_fn": _gemini_headers,
        "body_fn": _gemini_body,
        "parse_fn": _gemini_parse,
    },
    "deepseek": {
        "endpoint": "https://api.deepseek.com/chat/completions",
        "default_model": "deepseek-chat",
        "headers_fn": _deepseek_headers,
        "body_fn": _deepseek_body,
        "parse_fn": _deepseek_parse,
    },
    "jev": {
        # TypeSafe AI Jev (System 1 fast decision, judging, and routing model)
        "endpoint": "https://api.typesafe.ai/v1/chat/completions",
        "default_model": "typesafe/jev",
        "headers_fn": _openai_headers,
        "body_fn": _openai_body,
        "parse_fn": _openai_parse,
    },
    "laya": {
        # Convai Innovations Laya (System 1 fast local decision & classifier model)
        "endpoint": "http://127.0.0.1:8385/v1/chat/completions",
        "default_model": "convaiinnovations/laya",
        "headers_fn": _optional_bearer_headers,
        "body_fn": _openai_body,
        "parse_fn": _openai_parse,
    },
    "groq": {
        "endpoint": "https://api.groq.com/openai/v1/chat/completions",
        "default_model": "llama-3.3-70b-versatile",
        "headers_fn": _openai_headers,
        "body_fn": _openai_body,
        "parse_fn": _openai_parse,
    },
    "mistral": {
        "endpoint": "https://api.mistral.ai/v1/chat/completions",
        "default_model": "mistral-small-latest",
        "headers_fn": _openai_headers,
        "body_fn": _openai_body,
        "parse_fn": _openai_parse,
    },
    "openrouter": {
        "endpoint": "https://openrouter.ai/api/v1/chat/completions",
        "default_model": "typesafe/jev",
        "headers_fn": _openai_headers,
        "body_fn": _openai_body,
        "parse_fn": _openai_parse,
    },
    "ollama": {
        "endpoint": "http://localhost:11434/v1/chat/completions",
        "default_model": "llama3",
        "headers_fn": _optional_bearer_headers,
        "body_fn": _openai_body,
        "parse_fn": _openai_parse,
    },
    "together": {
        "endpoint": "https://api.together.xyz/v1/chat/completions",
        "default_model": "meta-llama/Llama-3.3-70B-Instruct-Turbo",
        "headers_fn": _openai_headers,
        "body_fn": _openai_body,
        "parse_fn": _openai_parse,
    },
    "qwen": {
        "endpoint": "https://dashscope.aliyuncs.com/compatible-mode/v1/chat/completions",
        "default_model": "qwen-plus",
        "headers_fn": _openai_headers,
        "body_fn": _openai_body,
        "parse_fn": _openai_parse,
    },
    "glm": {
        "endpoint": "https://open.bigmodel.cn/api/paas/v4/chat/completions",
        "default_model": "glm-4-flash",
        "headers_fn": _openai_headers,
        "body_fn": _openai_body,
        "parse_fn": _openai_parse,
    },
    "kimi": {
        "endpoint": "https://api.moonshot.cn/v1/chat/completions",
        "default_model": "moonshot-v1-8k",
        "headers_fn": _openai_headers,
        "body_fn": _openai_body,
        "parse_fn": _openai_parse,
    },
}

File: oma-ts/src/providers/http-providers.ts
/**
 * OMA HTTP Providers - raw HTTP connectors for each subscription.
 *
 * No SDK dependencies. Each provider is defined as:
 *   - endpoint URL
 *   - default model
 *   - headers builder
 *   - request body builder
 *   - response parser
 *
 * This is the side-channel: we talk to every provider the same way,
 * using their public HTTP APIs directly.
 */

import { Provider, type ProviderResponse, type Message, type RateLimiter, ErrorClass } from './base.js';

interface ParsedResponse {
  text: string;
  tokens_in: number;
  tokens_out: number;
  model?: string;
}

type HeadersFn = (apiKey: string) => Record<string, string>;
type BodyFn = (opts: { messages: Message[]; system?: string; model: string; max_tokens: number; temperature: number }) => unknown;
type ParseFn = (raw: Record<string, unknown>) => ParsedResponse;

export interface ProviderConfig {
  endpoint: string;
  default_model: string;
  headers_fn: HeadersFn;
  body_fn: BodyFn;
  parse_fn: ParseFn;
}

async function httpPost(url: string, headers: Record<string, string>, body: unknown, timeoutMs: number): Promise<Record<string, unknown>> {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);

  try {
    const resp = await fetch(url, {
      method: 'POST',
      headers,
      body: JSON.stringify(body),
      signal: controller.signal,
    });
    const raw = await resp.json() as Record<string, unknown>;
    if (!resp.ok) {
      throw Object.assign(new Error(`HTTP ${resp.status}: ${JSON.stringify(raw).slice(0, 500)}`), { status: resp.status });
    }
    return raw;
  } finally {
    clearTimeout(timer);
  }
}

export class HTTPProvider extends Provider {
  endpoint: string;
  model: string;
  private headersFn: HeadersFn;
  private bodyFn: BodyFn;
  private parseFn: ParseFn;

  constructor(opts: {
    name: string;
    api_key: string;
    endpoint: string;
    model: string;
    headers_fn: HeadersFn;
    body_fn: BodyFn;
    parse_fn: ParseFn;
  }) {
    super(opts.api_key);
    this.name = opts.name;
    this.endpoint = opts.endpoint;
    this.model = opts.model;
    this.headersFn = opts.headers_fn;
    this.bodyFn = opts.body_fn;
    this.parseFn = opts.parse_fn;
  }

  async complete(
    messages: Message[],
    system?: string,
    maxTokens = 4096,
    temperature = 0.3,
    opts?: Record<string, unknown>,
  ): Promise<ProviderResponse> {
    await this.rateLimiter.waitIfNeeded(maxTokens);

    const headers = this.headersFn(this.apiKey!);
    const model = (opts?.model as string) || this.model;
    const body = this.bodyFn({ messages, system, model, max_tokens: maxTokens, temperature });

    let endpoint = this.endpoint;
    // gemini needs model in URL and API key as query param
    if (this.name === 'gemini' && endpoint.includes('{model}')) {
      endpoint = endpoint.replace('{model}', model) + `?key=${this.apiKey}`;
    }

    const t0 = performance.now();
    try {
      const raw = await httpPost(endpoint, headers, body, this.timeout);
      const latency = performance.now() - t0;
      const result = this.parseFn(raw);
      this.rateLimiter.recordTokens(result.tokens_in + result.tokens_out);
      this.onSuccess();

      return {
        text: result.text,
        tokens_in: result.tokens_in,
        tokens_out: result.tokens_out,
        model: result.model || this.model,
        provider: this.name,
        latency_ms: latency,
        raw,
      };
    } catch (e) {
      const latency = performance.now() - t0;
      const errMsg = e instanceof Error ? e.message : String(e);
      return {
        text: '',
        tokens_in: 0,
        tokens_out: 0,
        model: this.model,
        provider: this.name,
        latency_ms: latency,
        error: errMsg.slice(0, 500),
        error_class: this.classifyError(e),
      };
    }
  }

  countTokens(text: string): number {
    return Math.ceil(text.length / 4);
  }
}

// ---- provider configs ----

function claudeHeaders(apiKey: string): Record<string, string> {
  return {
    'Content-Type': 'application/json',
    'x-api-key': apiKey,
    'anthropic-version': '2023-06-01',
  };
}

function claudeBody({ messages, system, model, max_tokens, temperature }: { messages: Message[]; system?: string; model: string; max_tokens: number; temperature: number }): unknown {
  const body: Record<string, unknown> = { model, max_tokens, temperature, messages };
  if (system) body.system = system;
  return body;
}

function claudeParse(raw: Record<string, unknown>): ParsedResponse {
  const content = raw.content as Array<Record<string, string>> | undefined;
  const text = content?.[0]?.text ?? '';
  const usage = raw.usage as Record<string, number> | undefined;
  return {
    text,
    tokens_in: usage?.input_tokens ?? 0,
    tokens_out: usage?.output_tokens ?? 0,
    model: raw.model as string | undefined,
  };
}

function openaiHeaders(apiKey: string): Record<string, string> {
  return {
    'Content-Type': 'application/json',
    'Authorization': `Bearer ${apiKey}`,
  };
}

function openaiBody({ messages, system, model, max_tokens, temperature }: { messages: Message[]; system?: string; model: string; max_tokens: number; temperature: number }): unknown {
  const msgs = [...messages];
  if (system) msgs.unshift({ role: 'system', content: system });
  return { model, max_tokens, temperature, messages: msgs };
}

function openaiParse(raw: Record<string, unknown>): ParsedResponse {
  const choices = raw.choices as Array<Record<string, unknown>> | undefined;
  const msg = choices?.[0]?.message as Record<string, string> | undefined;
  const text = msg?.content ?? '';
  const usage = raw.usage as Record<string, number> | undefined;
  return {
    text,
    tokens_in: usage?.prompt_tokens ?? 0,
    tokens_out: usage?.completion_tokens ?? 0,
    model: raw.model as string | undefined,
  };
}

function openaiResponsesBody({ messages, system, model, max_tokens }: { messages: Message[]; system?: string; model: string; max_tokens: number; temperature: number }): unknown {
  // GPT reasoning models reject temperature unless reasoning is disabled.
  // OMA manages context itself, so do not persist requests at the provider.
  const body: Record<string, unknown> = {
    model,
    input: [...messages],
    max_output_tokens: max_tokens,
    store: false,
  };
  if (system) body.instructions = system;
  return body;
}

function openaiResponsesParse(raw: Record<string, unknown>): ParsedResponse {
  const output = raw.output as Array<Record<string, unknown>> | undefined;
  const text = (output ?? [])
    .filter(item => item.type === 'message')
    .flatMap(item => (item.content as Array<Record<string, unknown>> | undefined) ?? [])
    .filter(content => content.type === 'output_text')
    .map(content => content.text as string | undefined)
    .filter((part): part is string => typeof part === 'string')
    .join('');
  const usage = raw.usage as Record<string, number> | undefined;
  return {
    text,
    tokens_in: usage?.input_tokens ?? 0,
    tokens_out: usage?.output_tokens ?? 0,
    model: raw.model as string | undefined,
  };
}

function geminiHeaders(_apiKey: string): Record<string, string> {
  return { 'Content-Type': 'application/json' };
}

function geminiBody({ messages, system, model, max_tokens, temperature }: { messages: Message[]; system?: string; model: string; max_tokens: number; temperature: number }): unknown {
  const contents = messages.map(m => ({
    role: m.role === 'user' ? 'user' : 'model',
    parts: [{ text: m.content }],
  }));
  const body: Record<string, unknown> = {
    contents,
    generationConfig: { maxOutputTokens: max_tokens, temperature },
  };
  if (system) body.systemInstruction = { parts: [{ text: system }] };
  return body;
}

function geminiParse(raw: Record<string, unknown>): ParsedResponse {
  const candidates = raw.candidates as Array<Record<string, unknown>> | undefined;
  const parts = (candidates?.[0]?.content as Record<string, unknown>)?.parts as Array<Record<string, string>> | undefined;
  const text = parts?.[0]?.text ?? '';
  const usage = raw.usageMetadata as Record<string, number> | undefined;
  return {
    text,
    tokens_in: usage?.promptTokenCount ?? 0,
    tokens_out: usage?.candidatesTokenCount ?? 0,
  };
}

function optionalBearerHeaders(apiKey: string): Record<string, string> {
  const h: Record<string, string> = { 'Content-Type': 'application/json' };
  if (apiKey) h['Authorization'] = `Bearer ${apiKey}`;
  return h;
}

export const PROVIDER_CONFIGS: Record<string, ProviderConfig> = {
  claude: {
    endpoint: 'https://api.anthropic.com/v1/messages',
    default_model: 'claude-sonnet-4-20250514',
    headers_fn: claudeHeaders,
    body_fn: claudeBody,
    parse_fn: claudeParse,
  },
  chatgpt: {
    // OpenAI Responses is the native protocol; the configs below retain the
    // Chat Completions protocol for third-party OpenAI-compatible APIs.
    endpoint: 'https://api.openai.com/v1/responses',
    default_model: 'gpt-5.4',
    headers_fn: openaiHeaders,
    body_fn: openaiResponsesBody,
    parse_fn: openaiResponsesParse,
  },
  gemini: {
    endpoint: 'https://generativelanguage.googleapis.com/v1beta/models/{model}:generateContent',
    default_model: 'gemini-2.0-flash',
    headers_fn: geminiHeaders,
    body_fn: geminiBody,
    parse_fn: geminiParse,
  },
  deepseek: {
    endpoint: 'https://api.deepseek.com/chat/completions',
    default_model: 'deepseek-chat',
    headers_fn: openaiHeaders,
    body_fn: openaiBody,
    parse_fn: openaiParse,
  },
  jev: {
    endpoint: 'https://api.typesafe.ai/v1/chat/completions',
    default_model: 'typesafe/jev',
    headers_fn: openaiHeaders,
    body_fn: openaiBody,
    parse_fn: openaiParse,
  },
  laya: {
    endpoint: 'http://127.0.0.1:8385/v1/chat/completions',
    default_model: 'convaiinnovations/laya',
    headers_fn: optionalBearerHeaders,
    body_fn: openaiBody,
    parse_fn: openaiParse,
  },
  groq: {
    endpoint: 'https://api.groq.com/openai/v1/chat/completions',
    default_model: 'llama-3.3-70b-versatile',
    headers_fn: openaiHeaders,
    body_fn: openaiBody,
    parse_fn: openaiParse,
  },
  mistral: {
    endpoint: 'https://api.mistral.ai/v1/chat/completions',
    default_model: 'mistral-small-latest',
    headers_fn: openaiHeaders,
    body_fn: openaiBody,
    parse_fn: openaiParse,
  },
  openrouter: {
    endpoint: 'https://openrouter.ai/api/v1/chat/completions',
    default_model: 'typesafe/jev',
    headers_fn: openaiHeaders,
    body_fn: openaiBody,
    parse_fn: openaiParse,
  },
  ollama: {
    endpoint: 'http://localhost:11434/v1/chat/completions',
    default_model: 'llama3',
    headers_fn: optionalBearerHeaders,
    body_fn: openaiBody,
    parse_fn: openaiParse,
  },
  together: {
    endpoint: 'https://api.together.xyz/v1/chat/completions',
    default_model: 'meta-llama/Llama-3.3-70B-Instruct-Turbo',
    headers_fn: openaiHeaders,
    body_fn: openaiBody,
    parse_fn: openaiParse,
  },
  qwen: {
    endpoint: 'https://dashscope.aliyuncs.com/compatible-mode/v1/chat/completions',
    default_model: 'qwen-plus',
    headers_fn: openaiHeaders,
    body_fn: openaiBody,
    parse_fn: openaiParse,
  },
  glm: {
    endpoint: 'https://open.bigmodel.cn/api/paas/v4/chat/completions',
    default_model: 'glm-4-flash',
    headers_fn: openaiHeaders,
    body_fn: openaiBody,
    parse_fn: openaiParse,
  },
  kimi: {
    endpoint: 'https://api.moonshot.cn/v1/chat/completions',
    default_model: 'moonshot-v1-8k',
    headers_fn: openaiHeaders,
    body_fn: openaiBody,
    parse_fn: openaiParse,
  },
};
